File: apps/server/src/modules/stratz/steam.ts
const STEAM64_OFFSET = 76561197960265728n
const STEAM_ID_PATTERN = /^\d{17}$/
const MAX_ACCOUNT_ID = 0xffffffffn

export function normalizeSteamId(value: string): string {
  const normalized = value.trim()
  if (!/^\d+$/.test(normalized)) {
    throw new Error(`Invalid Steam ID: ${value}`)
  }

  const number = BigInt(normalized)
  if (normalized.length === 17) {
    const accountId = number - STEAM64_OFFSET
    if (accountId < 0n || accountId > MAX_ACCOUNT_ID) {
      throw new Error(`Steam ID out of range: ${value}`)
    }
    return number.toString()
  }

  if (number > MAX_ACCOUNT_ID) {
    throw new Error(`Account ID out of range: ${value}`)
  }
  return (STEAM64_OFFSET + number).toString()
}

export function steamIdToAccountId(steamId: string): number {
  if (!STEAM_ID_PATTERN.test(steamId)) {
    throw new Error(`Invalid Steam ID: ${steamId}`)
  }
  const accountId = BigInt(steamId) - STEAM64_OFFSET
  if (accountId < 0n || accountId > MAX_ACCOUNT_ID) {
    throw new Error(`Steam ID out of range: ${steamId}`)
  }
  return Number(accountId)
}

export function accountIdToSteamId(accountId: number): string {
  return (STEAM64_OFFSET + BigInt(accountId)).toString()
}

File: astrbot-plugin/steam.py
STEAM64_OFFSET = 76561197960265728
MAX_ACCOUNT_ID = 0xFFFFFFFF


def normalize_steam_id(value: str) -> tuple[str, int]:
    """将 Steam64 或 Dota 短 ID 统一转换为 Steam64 和 account ID。"""
    value = value.strip()
    if not value.isdigit():
        raise ValueError("ID 只能包含数字")

    number = int(value)
    if len(value) == 17:
        account_id = number - STEAM64_OFFSET
        if account_id < 0 or account_id > MAX_ACCOUNT_ID:
            raise ValueError("Steam64 ID 不在有效范围内")
    else:
        account_id = number
        if account_id > MAX_ACCOUNT_ID:
            raise ValueError("短 ID 不在有效范围内")

    return str(STEAM64_OFFSET + account_id), account_id
